### PR TITLE
Add support for OAuth2 authentication using Client Secret and Refresh…

### PR DESCRIPTION
**Pull request by arunkumarchacko**
… token

Merging https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/661 to 2.2.x branch. Cannot do cherry-pick since the affected code is significantly different in master and 2.2.x branch.

### CODE REVIEW

File: util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java [hunk 1]
```diff
@@ -315,6 +316,19 @@ private Credential getCredentialsFromSAParameters(List<String> scopes, HttpTrans
     return new GoogleCredentialWithRetry(builder, options.getTokenServerUrl());
   }
 
+  private Credential getOAuthCredential(HttpTransport transport, List<String> scopes)
```
Comment (singhravidutt) [inline util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java:319]: where is `transport` getting used?

File: util-hadoop/src/main/java/com/google/cloud/hadoop/util/HadoopCredentialConfiguration.java [hunk 2]
```diff
@@ -204,6 +216,14 @@ public static CredentialFactory getCredentialFactory(
             .setProxyPassword(
                 RedactedString.create(
                     PROXY_PASSWORD_SUFFIX.withPrefixes(keyPrefixes).getPassword(config)))
+            .setAuthClientId(
+                AUTH_CLIENT_ID_SUFFIX.withPrefixes(keyPrefixes).get(config, config::get))
+            .setAuthClientSecret(
+                RedactedString.create(
+                    AUTH_CLIENT_SECRET_SUFFIX.withPrefixes(keyPrefixes).getPassword(config)))
```
Comment (singhravidutt) [inline util-hadoop/src/main/java/com/google/cloud/hadoop/util/HadoopCredentialConfiguration.java:223]: isn't this already of type `RedactedString`? do we need `RedactedString.create`?

File: util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java [hunk 3]
```diff
@@ -439,4 +458,12 @@ private HttpTransport getTransport() throws IOException {
   void setTransport(HttpTransport transport) {
     this.transport = transport;
   }
+
+  private static boolean isOAuthCredentialsConfigured(CredentialOptions options) {
```
Comment (singhravidutt) [inline util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java:462]: This seems so weird. Should we create bug in backlog to update this with `auth.type` token, same as in `master branch?

File: util-hadoop/src/test/java/com/google/cloud/hadoop/util/HadoopCredentialConfigurationTest.java [hunk 4]
```diff
@@ -76,6 +80,9 @@ public class HadoopCredentialConfigurationTest {
           put(".auth.impersonation.service.account", null);
           put(".auth.impersonation.service.account.for.user.", ImmutableMap.of());
           put(".auth.impersonation.service.account.for.group.", ImmutableMap.of());
+          put(".auth.client.id", null);
+          put(".auth.client.secret", null);
+          put(".auth.refresh.token", null);
```
Comment (medb) [inline util-hadoop/src/test/java/com/google/cloud/hadoop/util/HadoopCredentialConfigurationTest.java:85]: nit: sort alphabetically

**Review by medb — APPROVED** [review]

File: util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java (hunk 1 above)
Comment (arunkumarchacko) [reply util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java:319]: in the log

**Pull request merged**